### PR TITLE
[zero] update zero context init with the updated test utils

### PR DESCRIPTION
**Pull request by feifeibear**
1. help polishing non_distributed_component_funcs. return a function to build the model and optim.
2. add rm_torch_payload_on_the_fly for ZeroInitContext. Do not rm the torch param payload when adapting to ShardedParamV2.

### CODE REVIEW

File: tests/components_to_test/repeated_computed_layer.py [hunk 1]
```diff
@@ -36,9 +36,15 @@ def generate(self):
 
 @non_distributed_component_funcs.register(name='repeated_computed_layers')
 def get_training_components():
-    model = NetWithRepeatedlyComputedLayers(checkpoint=True)
+
+    def model_builder(checkpoint=True):
+        return NetWithRepeatedlyComputedLayers()
```
Comment (FrankLeeeee) [inline tests/components_to_test/repeated_computed_layer.py:41]: `checkpoint` is not passed to `NetWithRepeatedlyComputedLayers`

Comment (feifeibear) [reply tests/components_to_test/repeated_computed_layer.py:41]: DONE

File: tests/components_to_test/repeated_computed_layer.py [hunk 2]
```diff
@@ -36,9 +36,15 @@ def generate(self):
...
     trainloader = DummyDataLoader()
     testloader = DummyDataLoader()
-    optim = torch.optim.Adam(model.parameters(), lr=0.001)
+
+    def optim_builder(model):
+        return torch.optim.Adam(model.parameters(), lr=0.001)
+
     criterion = torch.nn.CrossEntropyLoss()
-    return model, trainloader, testloader, optim, criterion
+    return model_builder, trainloader, testloader, optim_builder, criterion
```
Comment (FrankLeeeee) [inline tests/components_to_test/repeated_computed_layer.py:50]: I believe this change will cause `tests/test_engine/test_engine.py` to fail. Can you refactor the `test_engine` test case as well?

File: tests/components_to_test/resnet.py [hunk 3]
```diff
@@ -22,9 +22,15 @@ def get_cifar10_dataloader(train):
 
 @non_distributed_component_funcs.register(name='resnet18')
 def get_resnet_training_components():
-    model = resnet18(num_classes=10)
+
+    def model_builder(checkpoint=False):
+        return resnet18(num_classes=10)
```
Comment (FrankLeeeee) [inline tests/components_to_test/resnet.py:27]: Is `checkpoint` here only for api consistency?

Comment (feifeibear) [reply tests/components_to_test/resnet.py:27]: Yes...

File: tests/components_to_test/repeated_computed_layer.py (hunk 2 above)
Comment (feifeibear) [reply tests/components_to_test/repeated_computed_layer.py:50]: DONE

File: tests/components_to_test/resnet.py (hunk 3 above)
Comment (FrankLeeeee) [reply tests/components_to_test/resnet.py:27]: ok.

**Review by FrankLeeeee — APPROVED** [review]

File: colossalai/zero/shard_utils/base_shard_strategy.py [hunk 4]
```diff
@@ -7,21 +7,30 @@
 class BaseShardStrategy(ABC):
 
     def __init__(self, process_group: Optional[dist.ProcessGroup] = None) -> None:
+        """Abstract Shard Strategy. Use to shard a tensors on multiple GPUs.
+
+        Args:
+            process_group (Optional[dist.ProcessGroup], optional): the process group. Defaults to None.
+        """
```
Comment (FrankLeeeee) [inline colossalai/zero/shard_utils/base_shard_strategy.py:14]: This format is not consistent with the current sphinx docstring style, thus, ReadTheDocs will not be able to parse this docstring.

Comment (feifeibear) [reply colossalai/zero/shard_utils/base_shard_strategy.py:14]: delete comments

Comment (FrankLeeeee) [reply colossalai/zero/shard_utils/base_shard_strategy.py:14]: got it

**Pull request merged**